WT-4451 Add new WiredTiger timestamp APIs that use integral values

sulabhM opened the pull request.

agorrod commented on dist/api_data.py:
```diff
@@ -1229,6 +1229,16 @@ def __ge__(self, other):
         choices=['commit', 'first_commit', 'prepare', 'read']),
 ]),
 
+'WT_SESSION.query_timestamp_numeric' : Method([
```
I don't love duplicating the information in the descriptions - could you reference the non-numeric versions doc instead? Similarly for timestamp_transaction_numeric?

sulabhM replied:
done, I have also modified wiredtiger.in to remove duplication, please check that too

agorrod commented on dist/api_data.py:
```diff
@@ -1379,6 +1389,29 @@ def __ge__(self, other):
         for a transaction. See @ref transaction_timestamps'''),
 ]),
 
+'WT_SESSION.timestamp_transaction_numeric' : Method([
```
@sueloverso do you have a naming preference for these two new APIs? We need to distinguish them from their equivalents that parse the number out of the config string.

sueloverso replied:
I cannot say I love the naming. But I don't have anything better to suggest either.

sueloverso replied:
After thinking about this a bit, what about `timestamp_transaction_value`. Although they're not the same thing it kind of flows with `cursor->set_value` and makes sense. I like that word a bit better than `_numeric`.

agorrod commented on lang/python/wiredtiger.i:
```diff
@@ -652,6 +653,35 @@ OVERRIDE_METHOD(__wt_cursor, WT_CURSOR, search_near, (self))
 
+/* Handle returned wt_timestamp_t timestamps. */
+%typemap(in,numinputs=0) (wt_timestamp_t *timestamp) (uint64_t ts) { $1 = &ts; }
+%typemap(frearg) (wt_timestamp_t *timestamp) "";
```
@ddanderson would you please take a look at the swig part of this change?

sulabhM replied:
Ahh! I suspected so, there are other instances of `frearg` that I copied from and don't think we need as well. I will remove these.

ddanderson commented on lang/python/wiredtiger.i:
```diff
@@ -652,6 +653,35 @@ OVERRIDE_METHOD(__wt_cursor, WT_CURSOR, search_near, (self))
+%typemap(frearg) (wt_timestamp_t *timestamp) "";
+%typemap(argout) (wt_timestamp_t *timestamp) { $result = PyLong_FromUnsignedLongLong(*$1); }
+
+/* 
```
In this comment, I would say that any numeric type (int or long in Py2, or int in Py3) is allowed as input for a timestamp.

sulabhM replied:
done

sulabhM commented on test/suite/test_assert01.py:
```diff
@@ -62,8 +59,8 @@ def insert_check(self, uri, use_ts):
         # Commit with a timestamp
         self.session.begin_transaction()
-        self.session.timestamp_transaction(
-            'commit_timestamp=' + timestamp_str(self.count))
```
added test for both with scenarios for some of the existing ones.

sulabhM commented on lang/python/wiredtiger.i:
```diff
@@ -652,6 +653,35 @@ OVERRIDE_METHOD(__wt_cursor, WT_CURSOR, search_near, (self))
+#else
+        if (PyInt_CheckExact($input)) {
+                if (PyInt_AS_LONG($input) < 0)
+                        SWIG_exception(SWIG_OverflowError,"can't convert negative long to unsigned.");
```
done, I changed it to what python 3 returns.

sulabhM commented on lang/python/wiredtiger.i:
```diff
@@ -652,6 +653,35 @@ OVERRIDE_METHOD(__wt_cursor, WT_CURSOR, search_near, (self))
+                        SWIG_exception(SWIG_OverflowError,"can't convert negative long to unsigned.");
+                $1 = PyInt_AsUnsignedLongLongMask($input);
+        } else if (PyLong_CheckExact($input))
+                $1 = PyLong_AsUnsignedLongLong($input);
```
I addressed bad input for python 2 and added a test to check all the possible inputs

sulabhM commented:
@ddanderson , if time permits have a look at the changes I made. I will try to address Java tomorrow. Haven't run Java ever with WT before, so if I can't get it running soon, I will create a separate ticket

ddanderson commented:
@sulabhM, thanks, the new changes for SWIG and python tests look good to me.

sulabhM commented:
@sueloverso , this pull request is ready for your review

sulabhM commented on dist/api_data.py:
```diff
@@ -1356,54 +1366,54 @@ def __ge__(self, other):
         current oldest and stable timestamps.  See
-        @ref transaction_timestamps'''),
+        @ref transaction_timestamps''',
+        min='0'),
```
This particular portion of the documentation has been removed. Same for other comments for `min=0`

sulabhM commented on dist/api_data.py:
```diff
@@ -1557,6 +1567,23 @@ def __ge__(self, other):
+            'oldest','oldest_reader','pinned','recovery','stable']),
+]),
+
 'WT_CONNECTION.set_timestamp' : Method([
```
I thought in one of the meetings we concluded that timestamp_transaction and query_timestamp are the two APIs that are going to have numeric versions.
@agorrod , @michaelcahill, @bvpvamsikrishna do we need a numeric version of set_timestamp ?

agorrod replied:
I don't think we need a version of `set_timestamp` - I think of this numeric API as an optimization/simplification path rather than a replacement for the current API. `WT_CONNECTION::set_timestamp` doesn't feel like it'll gain significant benefit from the new numeric API, so I'm inclined to skip it for now. If MongoDB wants it in the future we can add it as part of a different ticket.

bvpvamsikrishna replied:
I would like to maintain consistency, we can either have the numeric implementation or not have the numeric implementation, but better to be consistent with both WT_CONNECTION.set_timestamp and WT_CONNECTION.query_timestamp. Current state WT_CONNECTION.query_timestamp is implemented but not WT_CONNECTION.set_timestamp.

agorrod replied:
OK. I don't feel super strongly about it, so am OK to add a new `WT_CONNECTION::set_timestamp`. There are other APIs that accept timestamps as well and I don't think we should add numeric versions for all of those.

sueloverso replied:
I am feeling a bit ambivalent about this entire change. I mostly agree with @bvpvamsikrishna that set and query kind of go together. But I think that @dgottlieb's comment is most telling. The usage of the API of `WT_CONNECTION::set_timestamp("oldest_timestamp=N,stable_timestamp=N")` now has to become two API calls, and that introduces delay between those calls and ordering issues.

GeertBosch replied:
I agree with Sue and Dan that we need to address the multiple TS issue. If we still will have cases where we have to use the text version, it doesn't make sense to optimize only the other call sites. I don't think the formatting is a serious perf issue but will admit that I don't know how much WT parsing adds. We recently added a new string formatting library to the server that makes generating this kind of string much simpler and faster.

agorrod replied:
I'm not advocating for this change here - but does it make a difference that MongoDB doesn't currently call any API while passing in two timestamps to the same call?

dgottlieb replied:
MongoDB does make calls with multiple timestamps on the WT_CONNECTION::set_timestamp method (though now that I'm seeing whats in master, we may be doing it wrong in one spot!).
On v4.0 that is correct: https://github.com/mongodb/mongo/blob/ddb5e3eb82bd8f22a384835923d7f20bfa479d24/src/mongo/db/storage/wiredtiger/wiredtiger_kv_engine.cpp#L1506
In master that I think is missing the (duplicated) values:
https://github.com/mongodb/mongo/blob/ae3df9b0cb1327e818dc45da6bb3f0a1feb7f539/src/mongo/db/storage/wiredtiger/wiredtiger_kv_engine.cpp#L1603
Another example in master: https://github.com/mongodb/mongo/blob/ae3df9b0cb1327e818dc45da6bb3f0a1feb7f539/src/mongo/db/storage/wiredtiger/wiredtiger_kv_engine.cpp#L1671

sulabhM commented on src/txn/txn_timestamp.c:
```diff
@@ -341,16 +341,30 @@ __txn_query_timestamp(
+ *	global transaction or the session's transaction.
+ */
+int
+__wt_txn_query_timestamp_hex(WT_SESSION_IMPL *session,
```
I changed this to __wt_txn_query_timestamp_hex_str

sulabhM commented:
@ddanderson , I created WT-4806 for java related changes. It looks like we need broader changes in Java code, since I found no reference to timestamps in the examples.

sueloverso commented:
While I will do a full review, this entire ticket came out of a discussion with @GeertBosch and @dgottlieb last summer. I think they should chime in on whether these API changes are helpful and more efficient for them. No point in cluttering the API if they're not useful or won't be used.

dgottlieb commented on src/include/wiredtiger.in:
```diff
@@ -1937,9 +1944,36 @@ struct __wt_session {
+	 * @configend
+	 * @errors
+	 */
+	int __F(timestamp_transaction_numeric)(
```
Apologies if I've remarked on the wrong method/line:

Suppose I have a transaction in a prepared state. How would I write the analogous code for:
```
session->timestamp_transaction(session, "durable_timestamp=20,commit_timestamp=10");
```

agorrod replied:
That would need to become two calls to the new API:
```
session->timestamp_transaction_numeric(session, 20, "set=durable_timestamp");
session->timestamp_transaction_numeric(session, 10, "set=commit_timestamp");
```

We intend to keep both versions of the API - I suspect there are cases where making a single call to WiredTiger is preferrable to passing in numeric timestamp arguments.

sueloverso commented on examples/c/ex_all.c:
```diff
@@ -954,7 +954,14 @@ transaction_ops(WT_SESSION *session_arg)
 	char timestamp_buf[2 * sizeof(uint64_t) + 1];
 
+	/*! [transaction timestamp numeric] */
+	/* 29 hex is 41 decimal */
```
The comment "29 hex" is out of place because there is no 0x29 anywhere.

sulabhM commented:
I am holding off addressing review comments until we decide if it is okay to make multiple calls to set timestamps with `timestamp_transaction_numeric()`.

I had a quick look over MongoDB code and it does NOT look like we set multiple timestamps together with `timetsamp_transaction()` as of today.

sueloverso commented:
Given @GeertBosch's comment and my own doubts, I recommend that we not move forward with this change. It is cluttering of the API, gives multiple ways to do the same thing (which feels unclean) and does not provide any real benefit. Others may disagree.

@sulabhM I do NOT think this effort was wasted. It sounded like a good idea in theory and it was only when visually seeing what the API and changes would look like did I start reconsidering whether we really wanted this.

sulabhM commented:
As per the discussion here, we are not proceeding with this change. I am closing and discarding it.
